grammar.js

```diff
@@ -1,169 +1,278 @@
-      ),
-    expr: $ => seq("{", $.simplexpr, "}"),
-
-    // I think lookahead regexes are unfortunately not supported in tree-sitter, thus use an external scanner instead
-    // unescaped_string_fragment: _ => repeat(/[^\"\\]+(?:(?=\${)||(?=\\))/),
-    // I think right now every single character is allowed as escape sequence
-    escape_sequence: _ => token.immediate(seq("\\", /./)),
-    string_interpolation_start: _ => "${",
-    string_interpolation_end: _ => "}",
-    string_interpolation: $ =>
-      seq(
-        $.string_interpolation_start,
-        $.simplexpr,
-        $.string_interpolation_end
-      ),
+    source_file: $ => repeat($.ast_block),
+
+    ast_block: $ => seq(
+      '(',
+      alias($.symbol, $.widget_definition),
+      repeat($.expression),
+      ')',
+    ),
+
+    expression: $ => choice(
+      $.list,
+      $.array,
+      $.keyword,
+      $.symbol,
+      $._literal,
+      $.block,
+    ),
 
-    // Simple expression parser
-    // TODO prec necessary for atoms?
     simplexpr: $ =>
       choice(
-        prec.left(10, $.bool_literal),
-        prec.left(10, $.num_literal),
-        prec.left(10, $.string),
-        prec.left(10, $.ident),
-        prec.left(10, seq("(", $.simplexpr, ")")),
+        $._literal,
+        $.symbol,
         $.json_array,
         $.json_object,
-        $.function_call,
         $.json_access,
         $.json_safe_access,
         $.json_dot_access,
         $.json_safe_dot_access,
+        $.function_call,
         $.unary_expression,
         $.binary_expression,
-        $.ternary_expression
-      ),
-    ident: _ => /[a-zA-Z_][a-zA-Z0-9_-]*/,
-    json_array: $ => prec.left(10, seq("[", commaSep($.simplexpr), "]")),
-    json_object: $ =>
-      prec.left(
-        10,
-        seq("{", commaSep(seq($.simplexpr, ":", $.simplexpr)), "}")
-      ),
-    function_call: $ =>
-      prec.right(
-        9,
-        seq($.ident, field("operator", "("), commaSep($.simplexpr), ")")
+        $.ternary_expression,
+        $.parenthesized_expression,
       ),
-    json_access: $ =>
-      prec.right(
-        9,
-        seq(
-          $.simplexpr,
-          field("operator", "["),
-          alias($.simplexpr, $.index),
-          "]"
-        )
-      ),
-    json_safe_access: $ =>
-      prec.right(
-        9,
-        seq(
-          $.simplexpr,
-          field("operator", "?."),
-          "[",
-          alias($.simplexpr, $.index),
-          "]"
-        )
+
+    block: $ => seq('{', $.simplexpr, '}'),
+
+    json_array: $ => seq(
+      '[',
+      commaSep($.simplexpr),
+      ']',
+    ),
+    json_object: $ => seq(
+      '{',
+      commaSep(seq($.simplexpr, ':', $.simplexpr)),
+      '}',
+    ),
+    json_access: $ => seq(
+      $.simplexpr,
+      '[',
+      $.simplexpr,
+      ']',
+    ),
+    json_safe_access: $ => seq(
+      $.simplexpr,
+      '?.',
+      '[',
+      $.simplexpr,
+      ']',
+    ),
+    json_dot_access: $ => seq(
```

Comment:
This and for all the other json access access rules, the precision is missing, I think without the precision the AST is wrong, as this is kind of a binary expression, and `*` will have a higher precision than `.` or `?.`.

Thinking about it, I think also the "json atoms" (`json_array` and `json_object`) need to keep the precision for unambiguous correct parsing, and for the sake of completeness, I would also keep the parenthesized_expression precision (although I think it would parse unambiguously without it as well).

Again, keeping close to the original parser will avoid headaches and bugs introduced unnecessarily, only if you're absolutely sure that something is unnecessary, you may remove it.

Reply:
Parenthesized is unnecessary, but I'm confused as to how this will have conflicts/issues as it doesn't need that level of precendence comparatively. Keeping it close to the original parser is fine, but there's no conflicts without it

Reply:
I haven't tested it, but think about something like:

```
foo+bar.baz.abc
```

I think the precedence is something like this without explicit precedence:

```
(foo+bar).baz.abc
```
 
while it should be, according to the real grammar:
```
foo+(bar.baz.abc)
```

You can maybe write a test for that.

I think something similar happens, if you combine an array with either `json_access` or `json_safe_access`

Reply:
oh interesting, thanks I'll add that back

Reply:
please also keep the json_array and json_object precedence, just to be on the safe side (see lalrpop parser, one higher than json_access, it's on the same level as the atoms, I think there may still might be cases that are parsed incorrectly)

Reply:
done